### PR TITLE
Mapping CSV Tool and EOLS General Transformations Code

### PR DESCRIPTION
**Pull request by quazi-h**
## Why

[Relevant ticket](https://broadinstitute.atlassian.net/browse/DSPDC-1808)

## This PR
Added a mapping transformation script to process schema csvs to json schemas and fragments
Added EOLS Fragments and general conditions to the dog aging schema.
Added EOLS pipeline, pipeline builder, transformation code, and unit tests for general conditions.

### CODE REVIEW

**Comment by quazi-broad** [comment]
Realized that the RedCap data dict has some gating logic for some of those fields, added notes on what needs to be refactored.

File: scripts/convert_mapping_to_schema.py [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+    }
+    schema = render_schema_fragment(TDR_FRAGMENT_GENERAL, general_fragment, table_name)
+    schema["columns"].insert(0,dog_id_field)
+    schema["table_fragments"] = list(table_fragments)
+
+    return schema
+
+
+def render_schemas(fragments: dict[str, list[JadeColumn]], table_name: str):
+    general_fragment = fragments.pop(TDR_FRAGMENT_GENERAL)
```
Comment (aherbst-broad) [inline scripts/convert_mapping_to_schema.py:93]: 🔧 I think this should handle the edge case where there is no general fragment ( i.e., everything is bucketed) with a default empty dict:

```suggestion
    general_fragment = fragments.pop(TDR_FRAGMENT_GENERAL, {})
```

File: scripts/convert_mapping_to_schema.py [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+    for fragment, fields in fragments.items():
+        schema = render_schema_fragment(fragment, fields, table_name)
+        rendered_json = json.dumps(schema, indent=4)
+        print(os.getcwd())
```
Comment (aherbst-broad) [inline scripts/convert_mapping_to_schema.py:97]: ⛏️ Unless this debug output is really needed remove it.

File: scripts/convert_mapping_to_schema.py [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+        with open(f"schema/{table_name}_{fragment}.fragment.json", "w") as f:
```
Comment (aherbst-broad) [inline scripts/convert_mapping_to_schema.py:98]: 🔧 We should parameterize the output path (ie consume via a cli arg and thread through)

File: scripts/convert_mapping_to_schema.py [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+            f.write(rendered_json)
+
+    general_schema = render_general_fragment(general_fragment, [f"{table_name}_{key}" for key in fragments.keys()], table_name)
+    with open(f"schema/{table_name}.table.json", "w") as f:
```
Comment (aherbst-broad) [inline scripts/convert_mapping_to_schema.py:102]: 🔧 Same comment about parameterizing the output path.

File: scripts/convert_mapping_to_schema.py [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+        f.write(json.dumps(general_schema, indent=4))
+
+
+def run(table_name: str, mapping_csv: str) -> None:
```
Comment (aherbst-broad) [inline scripts/convert_mapping_to_schema.py:106]: ⛏️ This could use a more descriptive name. Perhaps `render_jade_table_schemas`?

**Review by aherbst-broad — CHANGES_REQUESTED** [review]
Looking good, a couple of points that could use cleanup in the schema generation script.

**Review by aherbst-broad — APPROVED** [review]
lgtm

**Review by raaidbroad — APPROVED** [review]

File: dap-etl/transformation/src/main/scala/org/broadinstitute/monster/dap/eols/EolsTransformationPipelineBuilder.scala [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+      .readJsonLists(
+        ctx,
+        "Raw EOLS Records",
+        s"${args.inputPrefix}/records/*.json"
+      )
+
+    // Group by study ID (record number) and arm (address_year_month)
+    // to get the format: (studyId, arm_id, Iterable((fieldName, Iterable(value))))
+    // (study_id, arm_id, iterable())
+    rawRecords
```
[the remark this answers is not in the document]
Comment (quazi-broad) [reply dap-etl/transformation/src/main/scala/org/broadinstitute/monster/dap/eols/EolsTransformationPipelineBuilder.scala:47]: Added a new ticket to address this: [DSPDC-1817](https://broadinstitute.atlassian.net/browse/DSPDC-1817)

**Pull request merged**